aioreactive/transform.py

```diff
@@ -465,3 +466,71 @@ def factory(exn: Exception) -> AsyncObservable[_TSource]:
         return pipe(source, catch(factory))
 
     return _retry
+
+
+def _scan(
+    accumulator: Callable[[_TResult, _TSource], Union[Awaitable[_TResult], _TResult]],
+    initial: _TResult,
+    async_result: bool = False,
+) -> Callable[[AsyncObservable[_TSource]], AsyncObservable[_TResult]]:
+    def _scan_operator(obs: AsyncObservable[_TSource]) -> AsyncObservable[_TResult]:
+        async def subscribe_async(observer: AsyncObserver[_TResult]) -> AsyncDisposable:
+            disposable: AsyncDisposable = AsyncDisposable.empty()
+            current: _TResult = initial
+
+            async def anext(value: _TSource) -> None:
+                nonlocal current
+
+                try:
+                    intermediate = accumulator(current, value)
+                except Exception as ex:
+                    await observer.athrow(ex)
+                else:
+                    if async_result:
+                        if not iscoroutine(intermediate):
+                            raise AssertionError(
+                                "The accumulator did not return an asynchronous result."
+                            )
+                        current = await intermediate
+                    else:
+                        current = intermediate  # type: ignore
+
+                    await observer.asend(current)
+
+            async def athrow(exception: Exception) -> None:
+                await observer.athrow(exception)
+
+            async def aclose() -> None:
+                await observer.aclose()
+
+            await disposable.dispose_async()
+            disposable = await obs.subscribe_async(anext, athrow, aclose)
+
+            return disposable
+
+        return AsyncAnonymousObservable(subscribe_async)
+
+    return _scan_operator
+
+
+def scan(
+    accumulator: Callable[[_TResult, _TSource], _TResult],
+    initial: _TResult,
+) -> Callable[[AsyncObservable[_TSource]], AsyncObservable[_TResult]]:
+    return _scan(accumulator, initial)
```

Comment:
Then make the sync case use an async accumulator as well:
```suggestion
def scan(
    accumulator: Callable[[_TResult, _TSource], _TResult],
    initial: _TResult,
) -> Callable[[AsyncObservable[_TSource]], AsyncObservable[_TResult]]:
    async def async_accumulator(state: _TResult, current: _TSource) -> _TResult:
        return accumulator(state, current)

    return _scan(async_accumulator, initial)
```